openmdao/components/exec_comp.py

```diff
@@ -342,16 +367,33 @@ def compute_partials(self, inputs, partials):
 
                 for u in out_names:
                     jval = imag(uwrap[u] / self.complex_stepsize)
-                    if (u, param) not in partials:  # create the dict entry
-                        partials[(u, param)] = np.zeros((jval.size, psize))
-
-                    # set the column in the Jacobian entry
-                    partials[(u, param)][:, i] = jval.flat
+                    if psize > 1 and jval.size > 1:
+                        partials[(u, param)] = jval.flat
+                    else:
+                        partials[(u, param)] = jval
 
                 # restore old param value
-                if idx is None:
-                    pwrap[param] -= step
-                else:
+                pwrap[param] -= step
+            else:
+                for i, idx in enumerate(array_idx_iter(pwrap[param].shape)):
+                    # set a complex param value
+                    pwrap[param][idx] += step
+
+                    uwrap = _TmpDict(self._outputs, return_complex=True)
+
+                    # solve with complex param value
+                    self._residuals.set_const(0.0)
+                    self.compute(pwrap, uwrap)
+
+                    for u in out_names:
+                        jval = imag(uwrap[u] / self.complex_stepsize)
```

Comment:
If uwrap[u] is an ndarry, then multiplying by the scalar (1.0 / complex_stepsize) is way faster than division.

Reply:
fixed